Resource based authorization

maliming opened the pull request:
Resolves #236

Demo: https://github.com/maliming/AbpResourcePermission

https://github.com/user-attachments/assets/574f89c5-c7ec-4a3f-a48b-c367d5172010


## Work to be done

- [x] Implement `IResourcePermissionChecker` in the Permission Management module. This is just for checking permissions, not managing (setting) them. Defined in the framework, so any module can be independent of the Permission Management module, yet can check permissions. We can implement as NullResourcePermissionChecker (just throw not implemented exception for now) in the framework to not get DI exceptions. 
- [x] Create `IResourcePermissionManager` to manage (get/set) permissions for resources in the Permission Management module, like the current IPermissionManager service. Implement it in the same place.
- [x] Create an app service to provide application logic to manage resource permissions.
- [x] Define a resource permission management modal, which should be reusable, like the current permission management modal.

hikalkan committed:
Introduce IResourcePermissionChecker

hikalkan committed:
Merge branch 'dev' into resource-based-auth

hikalkan committed:
Added GetGrantedResourceKeysAsync

hikalkan committed:
Added summaries

hikalkan committed:
Reformat IPermissionManager

hikalkan committed:
Remove outdated comment

hikalkan committed:
Create IResourcePermissionManager.cs

maliming committed:
Add resource permission management interfaces and implementations

maliming committed:
Add resource-based permission to PermissionDefinition system.

maliming committed:
Add resource-based permission value providers

maliming committed:
Refactor permission type to resource name

maliming committed:
Merge branch 'dev' into resource-based-auth

maliming committed:
Implement resource-based permission management with new methods and classes

maliming committed:
Enhance resource permission management by adding provider name and key parameters to permission checks

maliming committed:
Refactor GetPermissionsAsync method to return MultiplePermissionGrantResult instead of IDictionary

maliming committed:
Implement `ResourcePermissionStore`.

maliming committed:
Add resource permission grant repositories for EF Core and MongoDB

maliming committed:
Implement `ResourcePermissionManager`.

maliming committed:
Add resource permission checker and store extension methods for entity permissions

maliming committed:
Add resource permission management features and update API methods

maliming committed:
Make constructor public in permission handler

maliming committed:
Refactor permission management module

maliming committed:
Refactor permission grant info types in role provider

maliming committed:
feat(permission-management): add resource permission management features

maliming commented:
hi @hikalkan 

Our final step is to get user and role info in the **permissions** module. 

I plan to add two services to the **framework** and implement them in the **Identity** module to get user and role infor.

What do you think?

<img width="2360" height="1744" alt="image" src="https://github.com/user-attachments/assets/8c896278-1955-4343-8f79-14dae8a6e3c2" />

maliming committed:
fix(permission-management): validate user ID parsing in role permission providers

maliming committed:
feat: Enhance user and role management with search capabilities and resource provider integration

maliming committed:
feat(permission-management): update resource permission management modal with dynamic provider key selection

maliming committed:
feat: Add SearchByIdsAsync methods for users and roles in Identity module

maliming committed:
feat(permission-management): update resource permission handling and improve resource key usage in identity role and user management

maliming committed:
Remove resource permissions modal and related actions from roles and users

maliming committed:
Remove resource permission management script from roles and users pages

maliming committed:
Add 'ResourcePermissions' to English UI resources

maliming committed:
Remove TestPermissionDefinitionProvider and update localization keys for resource management modals

maliming committed:
Merge branch 'dev' into resource-based-auth

maliming committed:
Enhance permission management and localization for resource permissions

maliming committed:
Add localization for User and Role resource permission provider keys across multiple languages

maliming committed:
Refactor permission retrieval methods and clean up unused model validation in ResourcePermissionManagementModal

maliming committed:
Add missing closing brackets in localization files for Czech and English

maliming committed:
Add resource management permissions and update authorization attributes in PermissionAppService

maliming committed:
Fix form validation trigger on provider key change in resource permission management modal

maliming committed:
Update modal footer buttons in permission management modals for consistency

maliming committed:
Add localization for "NoResourceProviderKeyLookupServiceFound" message and update ResourcePermissionManagementModal to conditionally display modal content based on resource provider availability

maliming committed:
Implement resource permissions management with new interfaces and extension methods

maliming committed:
Add resource permission management to RoleUpdateEventHandler

maliming committed:
Refactor ResourcePermissionGrant properties and index for improved clarity and structure

maliming committed:
Refactor permission group assertions in tests

maliming committed:
Refactor resource permission population logic

maliming committed:
Merge branch 'dev' into resource-based-auth

maliming committed:
Add "NoResourcePermissionFound" localization key and update related modals

maliming committed:
Add ResourcePermissionGrant cache invalidator

maliming committed:
Add resource permission value provider tests

maliming committed:
Add resource permission checker tests and update permission definitions

maliming committed:
Add resource permission tests to StaticPermissionDefinitionStore

maliming committed:
Prevent child permissions for resource permissions

maliming committed:
Add resource permission grant repository tests

maliming committed:
Implement resource permission deletion in Role and User event handlers

maliming committed:
Add resource permission manager extensions for roles and users

maliming committed:
Add new tests for user and role search methods

maliming committed:
Return empty list for empty filter in search methods

maliming committed:
Add `ResourcePermissionStore_Tests`

maliming committed:
Add unit tests for ResourcePermissionChecker

maliming committed:
Add tests for ResourcePermissionGrant cache behavior

maliming committed:
Support resource permissions in permission serialization

maliming committed:
Add resource permission management provider and tests

maliming committed:
Add pagination support to resource permission search methods

maliming committed:
Update pagination logic in permission modal

hikalkan committed:
Introduce IKeyedObject and unify object key access

Added the IKeyedObject interface to provide a standard way to retrieve object keys, including support for composite keys via KeyedObjectHelper. Updated IEntity, IEntityDto, and related DTOs to implement IKeyedObject, and refactored resource permission logic to use GetObjectKey for key retrieval. This change improves consistency and reliability in handling object keys across the framework.

hikalkan committed:
Add GetObjectKey method to MyEntity class

Introduced the GetObjectKey method to MyEntity, returning the string representation of the entity's Id. This may be used for scenarios requiring a unique key in string format.

hikalkan committed:
Add tests for KeyedObjectHelper composite key methods

Introduces unit tests for encoding and decoding composite keys in the KeyedObjectHelper class, verifying correct behavior for various input scenarios.

hikalkan committed:
Remove JetBrains.Annotations and improve formatting

Removed all usages of JetBrains.Annotations attributes and related imports from resource permission interfaces and context classes. Reformatted method signatures and constructors for better readability and consistency.

hikalkan commented on framework/src/Volo.Abp.Authorization.Abstractions/Volo/Abp/Authorization/Permissions/IPermissionDefinitionContext.cs:
```diff
@@ -46,4 +48,15 @@ PermissionGroupDefinition AddGroup(
+        MultiTenancySides multiTenancySide = MultiTenancySides.Both,
+        bool isEnabled = true);
+
+    PermissionDefinition? GetResourcePermissionOrNull([NotNull] string name);
```
Should also take a resource name. Otherwise, how to know which resource's permission is requested? Or did you design resource permissions with unique names in the system?

hikalkan commented on framework/src/Volo.Abp.Authorization.Abstractions/Volo/Abp/Authorization/Permissions/IPermissionDefinitionContext.cs:
```diff
@@ -46,4 +48,15 @@ PermissionGroupDefinition AddGroup(
+
+    PermissionDefinition? GetResourcePermissionOrNull([NotNull] string name);
+
+    void RemoveResourcePermission([NotNull] string name);
```
Should also take a resource name. Otherwise, how to know which resource's permission is requested? Or did you design resource permissions with unique names in the system?

hikalkan committed:
Make EntityResourcePermissionRequirementHandler

hikalkan committed:
Rename ResourcePermissionExtenstions class

The class ResourcePermissionExtenstions was renamed to KeyedObjectResourcePermissionExtenstions to better reflect its purpose and usage.

maliming committed:
Merge branch 'dev' into resource-based-auth

maliming committed:
Merge branch 'dev' into resource-based-auth

maliming committed:
Remove redundant permission management translations

maliming committed:
Remove permission definition provider and permissions classes

maliming committed:
feat: add management permission to resource permissions and update related logic

maliming committed:
Refactor permission group creation in test provider

maliming committed:
refactor: update resource permission methods to include resourceName parameter

maliming committed:
Add tests for duplicate and multiple resource permissions

maliming committed:
Rename ManagementPermission to ManagementPermissionName

maliming committed:
Move KeyedObjectResourcePermission to Abp.Authorization

maliming committed:
Refactor permission app service for clarity

maliming committed:
Enforce management permission checks for resource permissions

maliming committed:
Grant TestEntityManagementPermission in FakePermissionStore

maliming committed:
Add test data for TestEntityManagementPermission

hikalkan committed:
Merge branch 'dev' into resource-based-auth

hikalkan commented on framework/src/Volo.Abp.Authorization/Volo/Abp/Authorization/Permissions/Resources/ResourcePermissionChecker.cs:
```diff
@@ -0,0 +1,179 @@
+            return false;
+        }
+
+        if (!await PermissionChecker.IsGrantedAsync(claimsPrincipal, permission.ManagementPermissionName!))
```
This permission is to check if the current user can change the permission of another user/role for a resource.
But here, we are not managing a user's permission. We are just checking the permission.

Example: User "admin" grants permission "Editing" of resource "blog-post" for User "halil". In this case, we should check the management permission. But, after that, when user "halil" logs in and tries to edit the blog post, we should only check if user "halil" has permission to edit the post. Management permission is irrelevant here. Understand? To properly test that, use two separate users.

maliming replied:
https://github.com/abpframework/abp/pull/24184/commits/12b50b41c5b5d2dbcc6aaa003c63e38b152cfb3b

maliming committed:
Remove management permission check from resource permissions

maliming committed:
Update expected resource permissions count in test

hikalkan committed:
Merge branch 'dev' into resource-based-auth

hikalkan committed:
Minor refactor

maliming committed:
Optimize permission checks using IPermissionChecker

maliming committed:
Unify 'ResourcePermissions' translation to 'Permissions'

maliming committed:
Update permission localization strings for clarity

maliming committed:
Remove redundant resource and provider localization strings from multiple language files

maliming committed:
Update resource permission localization and modals

maliming committed:
Improve resource permission modal UX and fix heading

maliming committed:
Refactor role search methods to use names instead of IDs and update related components

maliming committed:
Sanitize HTML id attributes by removing dots

maliming committed:
Show resource display name in permission modal header

maliming committed:
Update notification message on resource delete

maliming committed:
Add providers to resource permission DTO and update usage

maliming committed:
Fix permission filter order in resource modal
